### PR TITLE
DOC: better argcheck bar

### LINKED ISSUE #13142: Cannot plot bar graph with dates: "TypeError: ufunc subtract cannot use operands with types dtype('<M8[ns]') and dtype('float64')"
### Bug report

**Bug summary**

Unable to plot bar graph. Plotting a line graph works fine.

**Code for reproduction**

```python
from boto3.dynamodb.conditions import Key, Attr
import pandas as pd
import json
import decimal 
import matplotlib.pyplot as plt
import matplotlib.dates as mdate
from datetime import datetime

def timestamp_to_datetime(raw):
    return datetime.fromtimestamp(raw/1000)

df = pd.DataFrame(table.query(KeyConditionExpression=Key('partition_key').eq(name))['Items'])
df['date'] = df['timestamp'].apply(timestamp_to_datetime) 

plt.figure(figsize=(20,10))
plt.bar(df['date'], df['activity'].values)
plt.show();
```

**Actual outcome**

```
TypeError                                 Traceback (most recent call last)
<ipython-input-13-f28db9680a8e> in <module>()
     26 
     27 plt.figure(figsize=(20,10))
---> 28 plt.bar(df['date'], df['activity'].values)

c:\python27\lib\site-packages\matplotlib\pyplot.pyc in bar(*args, **kwargs)
   2773                       mplDeprecation)
   2774     try:
-> 2775         ret = ax.bar(*args, **kwargs)
   2776     finally:
   2777         ax._hold = washold

c:\python27\lib\site-packages\matplotlib\__init__.pyc in inner(ax, *args, **kwargs)
   1865                         "the Matplotlib list!)" % (label_namer, func.__name__),
   1866                         RuntimeWarning, stacklevel=2)
-> 1867             return func(ax, *args, **kwargs)
   1868 
   1869         inner.__doc__ = _add_data_doc(inner.__doc__,

c:\python27\lib\site-packages\matplotlib\axes\_axes.pyc in bar(self, *args, **kwargs)
   2262         if align == 'center':
   2263             if orientation == 'vertical':
-> 2264                 left = x - width / 2
   2265                 bottom = y
   2266             elif orientation == 'horizontal':

TypeError: ufunc subtract cannot use operands with types dtype('<M8[ns]') and dtype('float64')
```

**Expected outcome**

Expected: a plot of the (date, activity) in bar graph form. Changing plt.bar to plt.plot, as shown below, works perfectly fine:
```
from boto3.dynamodb.conditions import Key, Attr
import pandas as pd
import json
import decimal 
import matplotlib.pyplot as plt
import matplotlib.dates as mdate
from datetime import datetime

def timestamp_to_datetime(raw):
    return datetime.fromtimestamp(raw/1000)

df = pd.DataFrame(table.query(KeyConditionExpression=Key('partition_key').eq(name))['Items'])
df['date'] = df['timestamp'].apply(timestamp_to_datetime) 

plt.figure(figsize=(20,10))
**plt.plot(df['date'], df['activity'].values)**
plt.show();
```

**Matplotlib version**

  * Operating system: Windows 10 
  * Matplotlib version: 2.2.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 2.7

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Closes #13142

This checks that `x` and `width` in bar can be subtracted and emits a more useful message than before.

The underlying issue is that width=0.8, which has no units, and in this case x has datetime64 units.

### Code

```python
import pandas as pd
import matplotlib.pyplot as plt
from datetime import datetime

def timestamp_to_datetime(raw):
    return datetime.fromtimestamp(raw/1000)

df = pd.DataFrame([{
    'timestamp': 1500854400000,
    'activity': 5},
{
    'timestamp': 1500940800000,
    'activity': 6
}])
df['date'] = df['timestamp'].apply(timestamp_to_datetime)

plt.figure(figsize=(20,10))
plt.bar(df['date'], df['activity'])
plt.show();
```
### Before:

```
Traceback (most recent call last):
  File "testBarDT.py", line 24, in <module>
    plt.bar(df['date'], df['activity'])
  File "/Users/jklymak/matplotlib/lib/matplotlib/pyplot.py", line 2438, in bar
    **({"data": data} if data is not None else {}), **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/__init__.py", line 1778, in inner
    return func(ax, *args, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_axes.py", line 2277, in bar
    left = x - width / 2
TypeError: ufunc subtract cannot use operands with types dtype('<M8[ns]') and dtype('float64')
```

### After

```
Traceback (most recent call last):
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_axes.py", line 2278, in bar
    left = x - width / 2
TypeError: ufunc subtract cannot use operands with types dtype('<M8[ns]') and dtype('float64')

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "testBarDT.py", line 18, in <module>
    plt.bar(df['date'], df['activity'])
  File "/Users/jklymak/matplotlib/lib/matplotlib/pyplot.py", line 2438, in bar
    **({"data": data} if data is not None else {}), **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/__init__.py", line 1778, in inner
    return func(ax, *args, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_axes.py", line 2282, in bar
    f'({width.dtype}) are incompatible') from e
ValueError: the dtypes of parameters x (datetime64[ns]) and width (float64) are incompatible
```

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -2274,10 +2274,22 @@ def bar(self, x, height, width=0.8, bottom=None, *, align="center",
         # left, bottom, width, height vectors
         if align == 'center':
             if orientation == 'vertical':
-                left = x - width / 2
+                try:
+                    left = x - width / 2
+                except TypeError as e:
+                    raise ValueError('x and width parameters have '
```
[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: Yep, thats better...

[the remark this answers is not in the document]
Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: ...well except we want to make it clear they are parameters.  See the new wording above.

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -2274,10 +2274,20 @@ def bar(self, x, height, width=0.8, bottom=None, *, align="center",
         # left, bottom, width, height vectors
         if align == 'center':
             if orientation == 'vertical':
-                left = x - width / 2
+                try:
+                    left = x - width / 2
+                except TypeError as e:
+                    raise ValueError('the dtypes of parameters x '
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: Should still be a TypeError? I don't see why you'd change the exception type tbh.
Also, sorry for the OCDness, but can you put the linebreaks at the same place for x and y? :)

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: Yeah, I would have done that - @timhoffm suggest value error, but maybe mistakenly?  

As for OCD, admitting that you have a problem is the first step :wink:

Comment (jklymak) [reply lib/matplotlib/axes/_axes.py]: Changed to a TypeError - if @timhoffm has a reason for it not to be, he can comment.  Aligned the docstrings, though "height" is a longer word the "width", so there is only so much I can do there :wink:

File: lib/matplotlib/axes/_axes.py [hunk 3]
```diff
@@ -2274,10 +2274,20 @@ def bar(self, x, height, width=0.8, bottom=None, *, align="center",
         # left, bottom, width, height vectors
         if align == 'center':
             if orientation == 'vertical':
-                left = x - width / 2
+                try:
+                    left = x - width / 2
+                except TypeError as e:
+                    raise TypeError(f'the dtypes of parameters x ({x.dtype}) '
+                                     f'and width ({width.dtype}) '
+                                     'are incompatible') from e
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: These are both indented by one extra character.

**Comment by NelleV** [comment]
I've fixed the pep8 errors directly. I'll merge once test pass!

Thanks @jklymak !

**Comment by jklymak** [comment]
Ooops, sorry missed fixing those.  Feel free to squash merge, or I can self-merge when CI passes.

**Comment by dstansby** [comment]
Just restarted the tests since they're fixed, anyone feel free to merge if/when they pass!

**Comment by NelleV** [comment]
Thanks!

**Pull request merged**